src/data/right_inv.lean

```diff
@@ -0,0 +1,132 @@
+
+In this file we implement the bundled version of a right inverse function. The final purpose of
+this type is to conviniently implement sections of bundles.
+
+## Implementation notes
+
+Let `E : B → Type*` be the collection of fibers of a bundle on `B`.
+In the case of sections of bundles we prove that the type of right inverses of `bundle.proj E` is
+equivalent to the type `Π x : B, E x`. Both implementations of sections have their advantages:
+`right_inv (proj E)` is better to talk about properties such as continuity, whereas `Π x : B, E x`
+works better with algebraic structures: it naturally inherits the algebraic structures present on
+the fibers, thanks to the `Π` instances. It is hence a good idea to implement both and write down
+the equivalence between them, but then we have to choose only one for our final implementation of
+sections. We chose `right_inv (proj E)` because it is more general (some of the results in the
+theory of sections are true for `right_inv f`, for a generic `f`), and it has a more natural
+coercion to a function `B → bundle.total_space E` which is finally what we want. Note that we could
+also register a coercion to `B → bundle.total_space E` from `Π x : B, E x`, and it is not a bad idea
+to implement things this way, but it is just slightly more clumsy: in the case of continuous
+sections, in the first case one can implement continuous sections as
+`structure continuous_section (f : α → β) extends right_inv f, C(β, α)`
+whereas in the second way one would have to write continuity manually; it is not a big deal but the
+first way is slightly more natural. Also note that to recover the value of the section on the fiber
+in the first way one can use `.2` whereas the second way would make use of coercions.
+
+-/
+
+/-- Bundled right inverse of a function. Note that here the `nolint` has a true mathematical
+meaning: the structure is inhabited iff the function is surjective. -/
+@[nolint has_inhabited_instance]
+structure right_inv {α: Type*} {β: Type*} (f : α → β) :=
+(to_fun : β → α)
+(right_inv' : f ∘ to_fun = id)
+
+namespace right_inv
+
+variables {α: Type*} {β: Type*} {f : α → β} {g h : right_inv f}
+
+instance : has_coe_to_fun (right_inv f) := ⟨_, right_inv.to_fun⟩
+
+@[simp] lemma to_fun_eq_coe (g : right_inv f) : g.to_fun = ⇑g := rfl
+
+lemma coe_injective (H : ⇑g = h) : g = h :=
+by { cases g, cases h, congr' }
+
+@[ext] theorem ext (H : ∀ a, g a = h a) : g = h :=
+coe_injective $ funext H
+
+lemma right_inv_def {f : α → β} (g : right_inv f) : f ∘ g = id := g.right_inv'
+
+end right_inv
+
+section bundle_sections
+
+/-! ## Section of bundles -/
+
+open bundle
+
+variables {B : Type*} {E : B → Type*}
+
+@[simp] lemma right_inv.right_inv_apply (f : right_inv (proj E)) (b : B) : proj E (f b) = b :=
+by { exact congr_fun f.right_inv_def b, }
+
+@[simp] lemma right_inv.fst_eq_id (f : right_inv (proj E)) (b : B) : (f b).fst = b :=
+by { have h : (f b).fst = (proj E) (f b) := rfl, rw [h, f.right_inv_apply] }
```

Comment:
This second lemma is the same as the first, as `proj = sigma.fst` is true by rfl. If you're going to use `fst`, why introduce `proj` in the first place?

Reply:
Oh I don't want to use `fst` but sometimes it appears and `simp` does not apply automatically the above lemma to it

Reply:
Perhaps if lean let's you, you should add a simp lemma that turns `sigma.fst` on your bundle type into `proj`

Reply:
Hmm, you have a simp lemma that turns `proj` into `sigma.fist` already in the other direction. I think you should decide what you want the canonical spelling to be, and then write all your lemmas in terms of it

Reply:
You're right I actually do not recall adding that `@[simp]` there, maybe Sébastien did, so I guess I will take `.fst` to be the normal form and I will remove the first lemma.

Reply:
> This second lemma is the same as the first, as `proj = sigma.fst` is true by rfl. If you're going to use `fst`, why introduce `proj` in the first place?

Then the answer to this is that `sigma.fst` does not let me specify `E` and in most cases I don't apply it anywhere so lean cannot guess it. Plus, it makes notation closer to standard maths. Maybe I should mark the definition as `reducible`?

Reply:
Either making `proj` reducible or `bundle` irreducible seems like the way to go to me

Reply:
Ok then I'd make `proj` reducible, although I don't have a strong opinion on this. @sgouezel ?

Reply:
@sgouezel what do you prefer here?

Reply:
Well I made reducible even if there's a problem with it I wrote about below